explainers/subresource-loading.md

```diff
@@ -0,0 +1,102 @@
+# Explainer: Subresource loading with Web Bundles
+
+Last updated: March 2020
+
+We propose a new approach to load a large number of resources efficiently using
+a format that allows multiple resources be be bundled, e.g.
+[Web Bundles](https://web.dev/web-bundles/).
+
+## Backgrounds
+
+- Loading many unbundled resources is still slower in 2020. We concluded that
+  [bundling was necessary in 2018](https://v8.dev/features/modules#bundle), and
+  our latest local measurement still suggests that.
+
+- Caching stories with JS bundlers (e.g. webpack) is tough. They are pretty good
+  tools but configuring them to work in an optimal way is tough, and sometimes
+  also conflicts with new requirements like dynamic bundling (e.g. small edit
+  with tree shaking could invalidate everything).
+
+- With JS bundlers, parsing and execution needs to wait for the full bytes to
+  come. Ideally loading multiple subresources should be able to utilize full
+  streaming and parallelization, while it's not possible if all resources are
+  bundled as one javascript. (For JS modules execution still needs to be waited
+  for the entire tree due to the current
+  [deterministic execution model](https://docs.google.com/document/d/1MJK0zigKbH4WFCKcHsWwFAzpU_DZppEAOpYJlIW7M7E/edit#heading=h.5652gd5ks5id))
+
+- Related issues: [#411](https://github.com/WICG/webpackage/issues/411),
+  [#526](https://github.com/WICG/webpackage/issues/526)
+
+## Proposal
+
+We propose a new `link rel=”webbundle”` notion [**1**] that lets the browser
+fetch bundled resources and attach them to the page. Such as:
+
+```html
+<link
+  rel="webbundle"
+  href="https://example.com/dir/subresources.wbn"
+  scope="https://example.org/scope/"
+/>
+```
+
+Both the `href` and the `scope` need to be the same-origin.
```

Comment:
I don't think they need to be the same-origin, and actually the example URLs are not the same-origin.
When I wrote [this example](https://docs.google.com/document/d/18fFrURT6xD1GnqJwCx2K3Z81Te9Iv-r2ZsCkOvDH_04/edit#heading=h.irqypigubog7), my idea was:
  - If the page is controlled by a service worker, the service worker can intercept any requests from the page and can respond with arbitrary resources.
  - So, it is safe (=no difference from service worker from the security point of view) even if they are not the same origin, when the page can access the web bundle in the CORS manner.

Reply:
Thanks.
@kinu, what do you think? I'm afraid that I misunderstood your suggestions.

I'm wondering what the explainer should mention here. We have 3 URLs:

1. The main document's URL
2. href
3. scope

I thought 2 and 3 need to be the same origin, however, it is okay that 1 is in a different origin.
It's a simple mistake I forgot to update example URLs.

Any suggestions? If we don't have a clear answer, I think it is okay to mention it as TBD.

Reply:
Yeah I suggested this but I remember I discussed that cross-origin seems okay too with you. I want to think through all the implications... for example it apparently complicates CSP story. But I agree that maybe we can just drop this for now.

Reply:
(^^I commented before I see Hayato's response) @hayatoito I was indeed thinking about 2 and 3 indeed, but thinking about this more it's a little unclear too. (As @horo-t mentioned SW could already override it)  I think I'm fine with just dropping this sentence.

Reply:
Thanks! Okay, let's just drop this for now.
Once we have a more clear idea, let's add something back.